### PR TITLE
Support design library symbols and colors

### PR DESCRIPTION
**Pull request by cronik**
Allow badges to use Jenkins symbols and color references. Update info/warn/error badges to use symbols by default.

Screenshot 2024-05-02 at 11 13 25 AM

### Testing done

Used `run-fast` to create badges used in the screenshots.

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
Thanks very much @cronik.  I like the change!

I've made some documentation updates while I was testing the pull request.  If you object to those changes, you are welcome to remove them or revert them.

The `color` parameter that has been added to the command is not optional even though it is documented as optional in the javadoc and needs to be optional for compatibility with existing installations.

I haven't done the research to understand why that parameter is mandatory, though I suspect it is due to the new constructor that was added to the BadgeAction class.  I believe that new parameters are usually added with DataBoundSetter rather than by adding a new constructor

**Comment by cronik** [comment]
Thanks for those enhancements. 

The reason for adding `color` as a constructor instead of a setter was down to the fact that all the sub classes of `AddBadgeStep` don't support the color parameter, but the docs generator would include it. So in the interest of keeping the changes to a min I went the constructor route. Another option might be to adjust the Step type hierarchy a bit to get everything to align properly. I'll test some stuff out, maybe it's not as much additional change as I had originally thought.

**Comment by cronik** [comment]
@MarkEWaite I managed to rework the PR to move the color parameter to a setter and only apply it to the generic `addBadge` step. Was able to revert the doc generator changes as result.

**Comment by MarkEWaite** [comment]
Thanks @cronik!  I saw that there was no online help for the `addBadge` parameters, so I added online help for them and for a few other items.

Since the plugin is up for adoption, this pull request would be a great one for you to use to adopt the plugin.  Are you interested in adopting the plugin?  I'm happy to continue as a reviewer and tester, since I use the badge plugins and I maintain the embeddable build status plugin.

**Review by MarkEWaite — APPROVED** [review]
Changes have passed all my interactive tests

File: src/main/java/com/jenkinsci/plugins/badge/dsl/AbstractAddBadgeStep.java
Comment (MarkEWaite) [inline src/main/java/com/jenkinsci/plugins/badge/dsl/AbstractAddBadgeStep.java:4]: I'm not sure that you want to assign copyright for this file to Serban Iordache.  I think that you can either remove the explicit copyright statement or add your additional copyright if the original source was derived from something that Serban Iordache created.
```suggestion
 * Copyright (c) 2004-2010, Sun Microsystems, Inc., Serban Iordache
```

**Comment by MarkEWaite** [comment]
The online help that I added is working as expected on the `addBadge` step for all parameters except the `color` parameter.  I don't understand why the color parameter behaves differently in the online help than the help for the `icon`, `id`, `link`, and `text` parameters.  Unfortunately, I've run out of ideas to explore.  I've left the source file for the `text` help in the pull request, even though it is not displayed to the user from the "Pipeline syntax" page.

I think that it is an improvement to have online help, even if the help is not displayed for the `color` parameter.  The `color` parameter is well documented in the README.

The online help is visible from the "/pipeline-syntax" page when the `addBaadge` entry is selected in the dropdown menu and the "?" icon is clicked that is next to each parameter.  That text will be extracted into the ["Pipeline steps reference" page](https://www.jenkins.io/doc/pipeline/steps/badge/) after this pull request is merged and released.

**Comment by MarkEWaite** [comment]
While testing the combined pull request:

* #133 

I discovered that `var(--yellow)` does not change the icon color.  There is an example in the README that says:

```
addBadge(icon: 'symbol-star', text: 'A star', color: 'var(--yellow)')
```

Can you help me identify the mistake I am making when using that example?

**Comment by MarkEWaite** [comment]
The `createSummary` behavior has changed from the 1.9.1 release to this pull request.  The word "span" is incorrectly inserted into the message when I use `appendText` as suggested in one of the examples.

## Before this change

When I use the following Pipeline:

```
def summary = createSummary(icon: 'green')
summary.appendText('Completed text added.', true)
summary.appendText(' Bold text added.', false, true, false, 'black')
```

before

## After this change

after

File: src/main/java/com/jenkinsci/plugins/badge/action/AbstractAction.java
Comment (strangelookingnerd) [inline src/main/java/com/jenkinsci/plugins/badge/action/AbstractAction.java:66]: Wondering why we should limit this feature to the ionicons plugin alone and not also enable https://github.com/jenkinsci/font-awesome-api-plugin or https://github.com/jenkinsci/custom-folder-icon-plugin/blob/main/src/main/java/jenkins/plugins/foldericon/Emojis.java (and potentially others).

Comment (MarkEWaite) [reply src/main/java/com/jenkinsci/plugins/badge/action/AbstractAction.java:66]: Thanks for the comment @strangelookingnerd !  That's a very interesting suggestion, though I think it is outside of the scope of this pull request.  This pull request is already making a significant improvement to the plugin.

The plugin does not have a current maintainer.  It needs to be adopted so that a current maintainer can review the pull request, test it, merge it, and release it.  I've been testing this pull request because I'm an interested user of the plugin and I maintain the embeddable build status plugin.

Are you interested in becoming a [maintainer of the plugin](https://www.jenkins.io/doc/developer/plugin-governance/adopt-a-plugin/)?

Are you interested in submitting the pull request to add the additional support?

I've been submitting changes to the pull request while I've been testing it.  If you adopt the plugin, you'll be able to do the same and help @cronik on the journey to the release of this pull request.

**Comment by MarkEWaite** [comment]
> The `createSummary` behavior has changed from the 1.9.1 release to this pull request. The word "span" is incorrectly inserted into the message when I use `appendText` as suggested in one of the examples.

Fixed in f3091b5f87adbc56b912ee4d641b08b98f51f451

File: src/main/java/com/jenkinsci/plugins/badge/action/AbstractAction.java
Comment (mPokornyETM) [reply src/main/java/com/jenkinsci/plugins/badge/action/AbstractAction.java:65]: This is very good idea, open new issue (enhancement) and I will try to implement it here. Thx

**Review by mPokornyETM — APPROVED** [review]
LGTM

**Comment by cronik** [comment]
@MarkEWaite Thanks for the collaboration on this PR!

> 
> I discovered that `var(--yellow)` does not change the icon color. There is an example in the README that says:
> 

One issue is that the color style is not set on `l:icon`. But after some additional testing and digging I realized that even if  `<l:icon style="${badgeStyle}"` is set, `l:icon` does not respect it when the icon path starts with `symbol-`. So I think some revised docs are in order to indicate that only Jenkins palette and semantic color names and classes are support for symbol icon references. For other icon references those css references should work as expected with the update I just pushed.

It should be noted that `var(--yellow)` works with `addShortText`. I updated the help and docs to make this more clear.

**Comment by cronik** [comment]
> Since the plugin is up for adoption, this pull request would be a great one for you to use to adopt the plugin. Are you interested in adopting the plugin? I'm happy to continue as a reviewer and tester, since I use the badge plugins and I maintain the embeddable build status plugin.

@MarkEWaite I'm not entirely sure what it means to adopt a plugin, but I suppose I could.

**Comment by MarkEWaite** [comment]
@strangelookingnerd thanks very much for adopting the plugin.  I'm willing to merge and release it if needed, but would prefer that you would do that as a new maintainer.  I'm happy to answer questions or provide other encouragement as needed.

**Comment by strangelookingnerd** [comment]
@MarkEWaite I took the freedom of adding my remark from https://github.com/jenkinsci/badge-plugin/pull/130#discussion_r1590258007 to this PR. 
This is by not appending _plugin-ionicons-api_ to the icon name (in case they start with _symbol-_) and rather expect the icon name to follow the pattern of __symbol__-_icon-name_ __plugin__-_plugin-name_ as noted in https://weekly.ci.jenkins.io/design-library/Symbols.
I also updated the docs and pipeline syntax help. Let me know what you think and I'll gladly merge and release this PR.

**Review by strangelookingnerd — APPROVED** [review]

**Comment by MarkEWaite** [comment]
> @MarkEWaite I took the freedom of adding my remark from [#130 (comment)](https://github.com/jenkinsci/badge-plugin/pull/130#discussion_r1590258007) to this PR. This is by not appending _plugin-ionicons-api_ to the icon name (in case they start with _symbol-_) and rather expect the icon name to follow the pattern of **symbol**-_icon-name_ **plugin**-_plugin-name_ as noted in https://weekly.ci.jenkins.io/design-library/Symbols. I also updated the docs and pipeline syntax help. Let me know what you think and I'll gladly merge and release this PR.

That looks great to me.  I made minor changes to the documentation.  I think it is ready to be squash merged and released.

Thanks @cronik for the pull request and thanks @strangelookingnerd for adopting the plugin

**Pull request merged**
